tests/test_helpers.py

```diff
@@ -494,6 +499,515 @@ def test_set_content_disposition_bad_param():
+            )
+
+            A1 = '%s:%s:%s' % (username, realm, password)
+            A2 = '%s:%s' % (request.method, auth_data['uri'])
+
+            secret = H(A1)
+            data = '%s:%s:%s:%s:%s' % (
+                auth_data['nonce'],
+                auth_data['nc'],
+                auth_data['cnonce'],
+                auth_data['qop'],
+                H(A2)
+            )
+
+            assert H('%s:%s' % (secret, data)) == auth_data['response']
+
+            return web.Response()
+
+        response = web.Response(status=401)
+        data = '%s:%s' % (time.ctime().encode(), 'secret')
+        nonce = hashlib.sha1(data.encode()).hexdigest()
+        pairs = ', '.join([
+            'realm="%s"' % realm,
+            'qop="auth"',
+            'algorithm=MD5',
+            'nonce="%s"' % nonce,
+        ])
+        response.headers['WWW-Authenticate'] = 'Digest %s' % pairs
+        return response
+
+    app = web.Application()
+    app.router.add_route('GET', '/', handler)
+    client = yield from test_client(app)
+
+    auth = helpers.DigestAuth(username, password, client)
+
+    resp = yield from auth.request('GET', '/')
+    assert 200 == resp.status
+
+
+@asyncio.coroutine
+def test_digest_auth_bad_algorithm(loop, test_client):
+    opaque = binascii.hexlify(os.urandom(16))
+    username = 'username'
+    password = 'password'
+
+    def H(x):
+        if isinstance(x, str):
+            x = x.encode()
+        return hashlib.md5(x).hexdigest()
+
+    @asyncio.coroutine
+    def handler(request):
+        realm = 'test@example.com'
+        if 'Authorization' in request.headers:
+            assert request.headers['Authorization'] == ''
+
+            return web.Response(status=401)
+
+        response = web.Response(status=401)
+        data = '%s:%s' % (time.ctime().encode(), 'secret')
+        nonce = hashlib.sha1(data.encode()).hexdigest()
+        pairs = ', '.join([
+            'realm="%s"' % realm,
+            'qop="auth"',
+            'algorithm="nonsense"',
+            'nonce="%s"' % nonce,
+            'opaque="%s"' % opaque,
+        ])
+        response.headers['WWW-Authenticate'] = 'Digest %s' % pairs
+        return response
+
+    app = web.Application()
+    app.router.add_route('GET', '/', handler)
+    client = yield from test_client(app)
+
+    auth = helpers.DigestAuth(username, password, client)
+
+    resp = yield from auth.request('GET', '/')
+    assert 401 == resp.status
+
+
+
+@asyncio.coroutine
+def test_digest_auth_no_algorithm(loop, test_client):
+    opaque = binascii.hexlify(os.urandom(16))
+    username = 'username'
+    password = 'password'
+
+    def H(x):
+        if isinstance(x, str):
+            x = x.encode()
+        return hashlib.md5(x).hexdigest()
+
+    @asyncio.coroutine
+    def handler(request):
+        realm = 'test@example.com'
+        if 'Authorization' in request.headers:
+            pattern = re.compile(r'digest ', flags=re.IGNORECASE)
```

Comment:
ok, tests are the same as the regular code - please, show it some love in the name of maintenance.